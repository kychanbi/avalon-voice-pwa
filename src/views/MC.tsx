import React, {
  EventHandler, useCallback, useContext, useEffect, useState,
} from 'react';
import {
  Box, Card, Fab, makeStyles, Typography,
} from '@material-ui/core';
import PlayArrowIcon from '@material-ui/icons/PlayArrow';
import StopIcon from '@material-ui/icons/Stop';
import { AllSetting, SettingContext } from '../state/playerSetting';
import {
  endSpeaking, speak,
} from '../utils/utils';

interface settingDisplayProps {
  gameSetting: AllSetting
}

const useStyles = makeStyles((theme) => ({
  mcCard: {
    padding: theme.spacing(2),
  },
  playBtnWrap: {
    width: '100%',
    padding: '1em 0',
    textAlign: 'center',
  },
}));

const CurrentSettingDisplay = ({ gameSetting }: settingDisplayProps) => {
  const {
    numberOfGood, numberOfEvil, totalNumberOfPlayer,
  } = gameSetting;
  return (
    <Box>
      <Typography variant="subtitle1">
        總人數:
        {' '}
        {totalNumberOfPlayer}
      </Typography>
      <Typography variant="subtitle1">
        好人:
        {' '}
        {numberOfGood}
        {' '}
        個
      </Typography>
      <Typography variant="subtitle1">
        壞人:
        {' '}
        {numberOfEvil}
        {' '}
        個
      </Typography>
    </Box>
  );
};
const generateScript = ({
  isLancelotPresent,
  isMordredPresent,
  isMorganaPresent,
  isOberonPresent,
  isPercivalPresent,
}: AllSetting): string[] => {
  // let numOtherEvil = numberOfEvil;
  // isMordredPresent && numOtherEvil--;
  // isOberonPresent && numOtherEvil--;
  // isMorganaPresent && numOtherEvil--;
  // isLancelotPresent && numOtherEvil--;

  // const script1 = '';
  // let scriptEvil = '所有人合埋眼，伸手握拳放係枱上，所有壞人開眼';
  let scriptEvil = '所有壞人擘大眼';
  if (isLancelotPresent && isOberonPresent) scriptEvil = `除【奧柏倫同蘭斯洛特】以外，${scriptEvil}，蘭斯洛特豎起手指公`;
  else if (isOberonPresent) scriptEvil = `除左【奧柏倫】以外，${scriptEvil}`;
  else if (isLancelotPresent) scriptEvil = `除左【蘭斯洛特】以外，${scriptEvil}，蘭斯洛特豎起手指公`;
  const scriptEvil2 = '所有壞人合埋眼';
  const scriptEvil3 = `${isMordredPresent ? '除左【莫德雷德】以外，' : ''}壞人豎起手指公。梅林擘大眼`;
  // const numOfEvilShown = numberOfEvil - (+isMordredPresent);
  // const scriptGood = '';
  // ，`總共有 ${numOfEvilShown} 個壞人，如果數目有錯，請出聲。`;
  const scriptGood2 = '梅林合埋眼，壞人收起手指公';
  const scriptGood3 = isPercivalPresent ? `【梅林】${isMorganaPresent ? '及【莫甘娜】' : ''} 豎起手指公，【派西維爾】擘大眼。
` : '';
  const scriptGood4 = isPercivalPresent ? `【梅林】${isMorganaPresent ? '及【莫甘娜】' : ''} 收起手指公，【派西維爾】合埋眼。` : '';
  const scriptGood5 = '所有人擘大眼';
  return [
    scriptEvil, scriptEvil2, scriptEvil3,
    scriptGood2, scriptGood3, scriptGood4, scriptGood5,
  ].filter((s) => !!s);
};

const addLineBreaksToScript = (script: string[]): (string | JSX.Element)[] => (script).map(
  // eslint-disable-next-line react/jsx-key
  (x) => [x, <br />],
).reduce(
  (ac, curr) => (ac.concat(curr)), [],
);

const PlayButton = (props: { playOnClick: EventHandler<any> }) => {
  const { playOnClick } = props;
  return (
    <div onClick={playOnClick} role="button" tabIndex={0} onKeyDown={playOnClick}>
      <Fab variant="extended" color="primary">
        <PlayArrowIcon />
        {' '}
        Click to Play Script
      </Fab>
    </div>
  );
};

const StopButton = (props: { stopPlaying: EventHandler<any> }) => {
  const { stopPlaying } = props;
  return (
    <div onClick={stopPlaying} role="button" tabIndex={0} onKeyDown={stopPlaying}>
      <Fab variant="extended" color="secondary">
        <StopIcon />
      </Fab>
    </div>
  );
};

const useMCPageHook = () => {
  // const synthesize = useSynthesize();

  const { allSetting } = useContext(SettingContext);
  const [scriptArr, setScriptText] = useState<string[]>([]);
  const [scriptDisplay, setScriptDisplay] = useState<(string | JSX.Element)[]>([]);
  useEffect(() => {
    const s = generateScript(allSetting);
    const t = addLineBreaksToScript(s);
    setScriptText(s);
    setScriptDisplay(t);
  }, [allSetting]);

  const playScript = useCallback(() => {
    endSpeaking();
    // const lastSentence = scriptArr.pop() as string;
    scriptArr.forEach((s) => {
      speak('五，四，三，二，一．', allSetting.countingRate);
      speak(s, allSetting.speakingRate);
      // await pauseSpeaking(5);
    });
    // console.log('lastSentence',lastSentence)
    // speak(lastSentence);
    // synthesize(scriptText.join(' '));
  }, [scriptArr]);
  return {
    scriptArr, allSetting, playScript, scriptDisplay,
  };
};

export const MCPage = () => {
  const classes = useStyles();
  const {
    scriptDisplay, allSetting, playScript,
  } = useMCPageHook();
  // TODO : add loading state when SpeechSynthesisVoice is not loaded
  return (
    <Card variant="outlined" className={classes.mcCard}>
      <CurrentSettingDisplay gameSetting={allSetting} />
      <Typography>
        {scriptDisplay}
      </Typography>
      <div className={classes.playBtnWrap}>
        <PlayButton playOnClick={playScript} />
      </div>
      <div className={classes.playBtnWrap}>
        <StopButton stopPlaying={endSpeaking} />
      </div>
    </Card>
  );
};
