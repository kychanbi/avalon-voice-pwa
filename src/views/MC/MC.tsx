import React, { useCallback, useContext, useEffect, useState } from 'react';
import {
  Card,
  CircularProgress,
  makeStyles,
  Typography,
} from '@material-ui/core';

import {
  calcGoodEvilNumber,
  GameMode,
  SettingContext,
} from '../../state/playerSetting';
import { checkIfIos, endSpeaking, speak } from '../../utils/utils';
import { generateAvalonScript, generateQuestScript } from './generateScript';
import { t } from '@lingui/macro';
import PlayButton from './components/PlayButton';
import StopButton from './components/StopButton';
import CurrentSettingDisplay from './components/CurrentSettingDisplay';

const useStyles = makeStyles((theme) => ({
  mcCard: {
    padding: theme.spacing(2),
    height: '100%',
  },
  playBtnWrap: {
    width: '100%',
    padding: '1em 0',
    textAlign: 'center',
  },
}));

const addLineBreaksToScript = (script: string[]): (string | JSX.Element)[] =>
  script
    .map(
      // eslint-disable-next-line react/jsx-key
      (x) => [x, <br />],
    )
    .reduce((ac, curr) => ac.concat(curr), []);

const useMCPageHook = () => {
  const { allSetting } = useContext(SettingContext);
  const { questCharacterSetting, avalonCharacterSetting, isNewbieMode } =
    allSetting;
  const [scriptArr, setScriptText] = useState<string[]>([]);
  const [scriptDisplay, setScriptDisplay] = useState<(string | JSX.Element)[]>(
    [],
  );
  const [areVoicesLoaded, setAreVoicesLoad] = useState(false);
  const { numberOfEvil } = calcGoodEvilNumber(
    avalonCharacterSetting.totalNumberOfPlayer,
  );
  useEffect(() => {
    const s =
      allSetting.gameMode === GameMode.Avalon
        ? generateAvalonScript({
            ...avalonCharacterSetting,
            numberOfEvil,
            isNewbieMode,
          })
        : generateQuestScript({
            ...questCharacterSetting,
          });
    const t = addLineBreaksToScript(s);
    setScriptText(s);
    setScriptDisplay(t);
  }, [allSetting]);

  const checkIfVoiceLoaded = useCallback(() => {
    const synth = window.speechSynthesis;
    const voices = synth.getVoices();
    if (voices.length > 0) {
      setAreVoicesLoad(true);
    }
  }, [setAreVoicesLoad]);

  // const { i18n } = useLingui();

  useEffect(() => {
    if (checkIfIos()) {
      checkIfVoiceLoaded();
    } else {
      window.speechSynthesis.onvoiceschanged = checkIfVoiceLoaded;
      checkIfVoiceLoaded();
    }
  }, []);
  const playScript = useCallback(() => {
    endSpeaking();
    scriptArr.forEach((s) => {
      speak(t`三，二，一．`, allSetting.countingRate, allSetting.language);
      speak(s, allSetting.speakingRate, allSetting.language);
    });
  }, [scriptArr, allSetting.language]);
  return {
    scriptArr,
    allSetting,
    playScript,
    scriptDisplay,
    areVoicesLoaded,
  };
};

export const MCPage = () => {
  const classes = useStyles();
  const { scriptDisplay, allSetting, playScript, areVoicesLoaded } =
    useMCPageHook();
  return (
    <Card variant="outlined" className={classes.mcCard}>
      <CurrentSettingDisplay gameSetting={allSetting} />
      <Typography>{scriptDisplay}</Typography>
      <div className={classes.playBtnWrap}>
        {areVoicesLoaded ? (
          <PlayButton playOnClick={playScript} />
        ) : (
          <CircularProgress />
        )}
      </div>
      <div className={classes.playBtnWrap}>
        <StopButton stopPlaying={endSpeaking} />
      </div>
    </Card>
  );
};
