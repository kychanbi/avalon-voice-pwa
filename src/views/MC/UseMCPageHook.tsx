import React, { useCallback, useContext, useEffect, useState } from 'react';
import {
  calcGoodEvilNumber,
  GameMode,
  SettingContext,
} from '../../state/playerSetting';
import { generateAvalonScript, generateQuestScript } from './generateScript';
import { checkIfIos } from '../../utils/utils';
import { endSpeaking, speak } from '../../utils/speakingUtils';
import { t } from '@lingui/macro';
import { useVoices } from '../../state/useVoices';

const addLineBreaksToScript = (script: string[]): (string | JSX.Element)[] =>
  script.map((x) => [x, <br />]).reduce((ac, curr) => ac.concat(curr), []);

export const useMCPageHook = () => {
  const { allSetting, selectedVoice, areVoicesLoaded } =
    useContext(SettingContext);
  const { questCharacterSetting, avalonCharacterSetting, isNewbieMode } =
    allSetting;
  const [scriptArr, setScriptText] = useState<string[]>([]);
  const [scriptDisplay, setScriptDisplay] = useState<(string | JSX.Element)[]>(
    [],
  );
  const { numberOfEvil } = calcGoodEvilNumber(
    avalonCharacterSetting.totalNumberOfPlayer,
  );
  useEffect(() => {
    const s =
      allSetting.gameMode === GameMode.Avalon
        ? generateAvalonScript({
            ...avalonCharacterSetting,
            numberOfEvil,
            isNewbieMode,
          })
        : generateQuestScript({
            ...questCharacterSetting,
          });
    const t = addLineBreaksToScript(s);
    setScriptText(s);
    setScriptDisplay(t);
  }, [allSetting]);

  // const checkIfVoiceLoaded = useCallback(() => {
  //   const synth = window.speechSynthesis;
  //   const voices = synth.getVoices();
  //   if (voices.length > 0) {
  //     setAreVoicesLoad(true);
  //   }
  // }, [setAreVoicesLoad]);

  // useEffect(() => {
  //   if (checkIfIos()) {
  //     checkIfVoiceLoaded();
  //   } else {
  //     //onvoiceschanged will trigger when the voices are acutally loaded
  //     window.speechSynthesis.onvoiceschanged = checkIfVoiceLoaded;
  //     checkIfVoiceLoaded();
  //   }
  // }, []);
  const playScript = useCallback(() => {
    endSpeaking();
    scriptArr.forEach((s) => {
      speak(t`三，二，一．`, allSetting.countingRate, selectedVoice);
      speak(s, allSetting.speakingRate, selectedVoice);
    });
  }, [scriptArr, allSetting.language, selectedVoice]);
  return {
    scriptArr,
    allSetting,
    playScript,
    scriptDisplay,
    areVoicesLoaded,
  };
};
