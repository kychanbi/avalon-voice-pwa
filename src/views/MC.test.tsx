import { generateAvalonScript } from './MC';

describe('Script generation testing (2 evils game)', () => {
  it('(0 special evils) script generated correctly', () => {
    const useLancelotAlternativeRules = false;
    const isLancelotPresent = false;
    const isMordredPresent = false;
    const isMorganaPresent = false;
    const isOberonPresent = false;
    const isPercivalPresent = false;
    const isNewbieMode = false;
    const numberOfEvil = 2;
    const scripts = JSON.stringify(generateAvalonScript({
      useLancelotAlternativeRules,
      isLancelotPresent,
      isMordredPresent,
      isMorganaPresent,
      isOberonPresent,
      isPercivalPresent,
      isNewbieMode,
      numberOfEvil,
    }));
    expect(scripts)
      .toMatchSnapshot();
  });

  it('(Mordred present) script generated correctly', () => {
    const useLancelotAlternativeRules = false;
    const isLancelotPresent = false;
    const isMordredPresent = true;
    const isMorganaPresent = false;
    const isOberonPresent = false;
    const isPercivalPresent = false;
    const isNewbieMode = false;
    const numberOfEvil = 2;
    const scripts = JSON.stringify(generateAvalonScript({
      useLancelotAlternativeRules,
      isLancelotPresent,
      isMordredPresent,
      isMorganaPresent,
      isOberonPresent,
      isPercivalPresent,
      isNewbieMode,
      numberOfEvil,
    }));
    expect(scripts)
      .toMatchSnapshot();
  });

  it('(Mordred & Oberon present) script generated correctly', () => {
    const useLancelotAlternativeRules = false;
    const isLancelotPresent = false;
    const isMordredPresent = true;
    const isMorganaPresent = false;
    const isOberonPresent = true;
    const isPercivalPresent = false;
    const isNewbieMode = false;
    const numberOfEvil = 2;
    const scripts = JSON.stringify(generateAvalonScript({
      useLancelotAlternativeRules,
      isLancelotPresent,
      isMordredPresent,
      isMorganaPresent,
      isOberonPresent,
      isPercivalPresent,
      isNewbieMode,
      numberOfEvil,
    }));
    expect(scripts)
      .toMatchSnapshot();
  });

  it('(Mordred & Oberon present) newbie on script generated correctly', () => {
    const useLancelotAlternativeRules = false;
    const isLancelotPresent = false;
    const isMordredPresent = true;
    const isMorganaPresent = false;
    const isOberonPresent = true;
    const isPercivalPresent = false;
    const isNewbieMode = false;
    const numberOfEvil = 2;
    const scripts = JSON.stringify(generateAvalonScript({
      useLancelotAlternativeRules,
      isLancelotPresent,
      isMordredPresent,
      isMorganaPresent,
      isOberonPresent,
      isPercivalPresent,
      isNewbieMode,
      numberOfEvil,
    }));
    expect(scripts)
      .toMatchSnapshot();
  });
});

describe('Script generation testing (3 evils game)', () => {
  it('(0 special evils) script generated correctly(newbie on)', () => {
    const useLancelotAlternativeRules = false;
    const isLancelotPresent = false;
    const isMordredPresent = false;
    const isMorganaPresent = false;
    const isOberonPresent = false;
    const isPercivalPresent = false;
    const isNewbieMode = true;
    const numberOfEvil = 3;
    const scripts = JSON.stringify(generateAvalonScript({
      useLancelotAlternativeRules,
      isLancelotPresent,
      isMordredPresent,
      isMorganaPresent,
      isOberonPresent,
      isPercivalPresent,
      isNewbieMode,
      numberOfEvil,
    }));
    expect(scripts)
      .toMatchSnapshot();
  });
  it('(Mordred present) script generated correctly(newbie on)', () => {
    const useLancelotAlternativeRules = false;
    const isLancelotPresent = false;
    const isMordredPresent = false;
    const isMorganaPresent = false;
    const isOberonPresent = false;
    const isPercivalPresent = false;
    const isNewbieMode = true;
    const numberOfEvil = 3;
    const scripts = JSON.stringify(generateAvalonScript({
      useLancelotAlternativeRules,
      isLancelotPresent,
      isMordredPresent,
      isMorganaPresent,
      isOberonPresent,
      isPercivalPresent,
      isNewbieMode,
      numberOfEvil,
    }));
    expect(scripts)
      .toMatchSnapshot();
  });
  it('(Mordred Morgana present) script generated correctly', () => {
    const useLancelotAlternativeRules = false;
    const isLancelotPresent = false;
    const isMordredPresent = true;
    const isMorganaPresent = true;
    const isOberonPresent = false;
    const isPercivalPresent = true;
    const isNewbieMode = false;
    const numberOfEvil = 3;
    const scripts = JSON.stringify(generateAvalonScript({
      useLancelotAlternativeRules,
      isLancelotPresent,
      isMordredPresent,
      isMorganaPresent,
      isOberonPresent,
      isPercivalPresent,
      isNewbieMode,
      numberOfEvil,
    }));
    expect(scripts)
      .toMatchSnapshot();
  });
  it('(Mordred Morgana Oberon present) script generated correctly', () => {
    const useLancelotAlternativeRules = false;
    const isLancelotPresent = false;
    const isMordredPresent = true;
    const isMorganaPresent = true;
    const isOberonPresent = true;
    const isPercivalPresent = true;
    const isNewbieMode = false;
    const numberOfEvil = 3;
    const scripts = JSON.stringify(generateAvalonScript({
      useLancelotAlternativeRules,
      isLancelotPresent,
      isMordredPresent,
      isMorganaPresent,
      isOberonPresent,
      isPercivalPresent,
      isNewbieMode,
      numberOfEvil,
    }));
    expect(scripts)
      .toMatchSnapshot();
  });
  it('(Mordred Morgana Lancelot present) script generated correctly', () => {
    const useLancelotAlternativeRules = false;
    const isLancelotPresent = true;
    const isMordredPresent = true;
    const isMorganaPresent = true;
    const isOberonPresent = false;
    const isPercivalPresent = true;
    const isNewbieMode = false;
    const numberOfEvil = 3;
    const scripts = JSON.stringify(generateAvalonScript({
      useLancelotAlternativeRules,
      isLancelotPresent,
      isMordredPresent,
      isMorganaPresent,
      isOberonPresent,
      isPercivalPresent,
      isNewbieMode,
      numberOfEvil,
    }));
    expect(scripts)
      .toMatchSnapshot();
  });
});
