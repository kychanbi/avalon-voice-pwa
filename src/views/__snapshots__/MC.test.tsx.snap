// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Script generation testing (2 evils game) (0 special evils) script generated correctly 1`] = `"[\\"所有壞人擘大眼\\",\\"所有壞人合埋眼\\",\\"壞人豎起手指公，梅林擘大眼\\",\\"梅林合埋眼，壞人收起手指公\\",\\"所有人擘大眼\\"]"`;

exports[`Script generation testing (2 evils game) (Mordred & Oberon present) newbie on script generated correctly 1`] = `"[\\"除左【奧柏倫】以外，所有壞人擘大眼\\",\\"所有壞人合埋眼\\",\\"除左【莫德雷德】以外，壞人豎起手指公，梅林擘大眼\\",\\"梅林合埋眼，壞人收起手指公\\",\\"所有人擘大眼\\"]"`;

exports[`Script generation testing (2 evils game) (Mordred & Oberon present) script generated correctly 1`] = `"[\\"除左【奧柏倫】以外，所有壞人擘大眼\\",\\"所有壞人合埋眼\\",\\"除左【莫德雷德】以外，壞人豎起手指公，梅林擘大眼\\",\\"梅林合埋眼，壞人收起手指公\\",\\"所有人擘大眼\\"]"`;

exports[`Script generation testing (2 evils game) (Mordred present) script generated correctly 1`] = `"[\\"所有壞人擘大眼\\",\\"所有壞人合埋眼\\",\\"除左【莫德雷德】以外，壞人豎起手指公，梅林擘大眼\\",\\"梅林合埋眼，壞人收起手指公\\",\\"所有人擘大眼\\"]"`;

exports[`Script generation testing (3 evils game) (0 special evils) script generated correctly(newbie on) 1`] = `"[\\"所有壞人擘大眼，總共有3個壞人開眼\\",\\"所有壞人合埋眼\\",\\"壞人豎起手指公，梅林擘大眼，總共有3隻手指公\\",\\"梅林合埋眼，壞人收起手指公\\",\\"所有人擘大眼\\"]"`;

exports[`Script generation testing (3 evils game) (Mordred Morgana Lancelot present) script generated correctly 1`] = `"[\\"除左【壞蘭斯洛特】以外，所有壞人擘大眼，壞蘭斯洛特豎起手指公\\",\\"所有壞人合埋眼\\",\\"除左【莫德雷德】以外，壞人豎起手指公，梅林擘大眼\\",\\"梅林合埋眼，壞人收起手指公\\",\\"【梅林】及【莫甘娜】豎起手指公，【派西維爾】擘大眼。\\",\\"【梅林】及【莫甘娜】收起手指公，【派西維爾】合埋眼。\\",\\"所有人擘大眼\\"]"`;

exports[`Script generation testing (3 evils game) (Mordred Morgana Oberon present) script generated correctly 1`] = `"[\\"除左【奧柏倫】以外，所有壞人擘大眼\\",\\"所有壞人合埋眼\\",\\"除左【莫德雷德】以外，壞人豎起手指公，梅林擘大眼\\",\\"梅林合埋眼，壞人收起手指公\\",\\"【梅林】及【莫甘娜】豎起手指公，【派西維爾】擘大眼。\\",\\"【梅林】及【莫甘娜】收起手指公，【派西維爾】合埋眼。\\",\\"所有人擘大眼\\"]"`;

exports[`Script generation testing (3 evils game) (Mordred Morgana present) script generated correctly 1`] = `"[\\"所有壞人擘大眼\\",\\"所有壞人合埋眼\\",\\"除左【莫德雷德】以外，壞人豎起手指公，梅林擘大眼\\",\\"梅林合埋眼，壞人收起手指公\\",\\"【梅林】及【莫甘娜】豎起手指公，【派西維爾】擘大眼。\\",\\"【梅林】及【莫甘娜】收起手指公，【派西維爾】合埋眼。\\",\\"所有人擘大眼\\"]"`;

exports[`Script generation testing (3 evils game) (Mordred present) script generated correctly(newbie on) 1`] = `"[\\"所有壞人擘大眼，總共有3個壞人開眼\\",\\"所有壞人合埋眼\\",\\"壞人豎起手指公，梅林擘大眼，總共有3隻手指公\\",\\"梅林合埋眼，壞人收起手指公\\",\\"所有人擘大眼\\"]"`;
